prov/tcp/src/tcpx_ep.c

```diff
@@ -627,8 +627,12 @@ static int tcpx_ep_close(struct fid *fid)
 					 &ep->util_ep.ep_fid.fid);
 	}
 
-	if (ep->fid && ep->fid->fclass == TCPX_CLASS_CM)
+	if (ep->state == TCPX_RCVD_REQ && ep->fid) {
+		assert(ep->fid->fclass == FI_CLASS_CONNREQ);
+		free(ep->fid);
+	} else if (ep->fid && ep->fid->fclass == TCPX_CLASS_CM) {
 		tcpx_free_cm_ctx(ep->cm_ctx);
+	}
```

Comment:
I think we'll need this to be:
```
if (ep->state != TCPX_RCVD_REQ && ep->fid) {
    assert(ep->fid->fclass == FI_CLASS_CM);
    tcpx_free_cm_ctx(ep->cm_ctx);
}
```
see comment below.

Though, we're really trying to free ep->cm_ctx.  Could we just check that directly instead?

Reply:
Your suggestion makes sense. Let me update the code and re-upload.

Reply:
PR was updated with your suggested and commit message was updated accordingly. Thanks again!

Reply:
Are you able to confirm this fixes the ASAN report?

Reply:
I haven't tried to reproduce the bug for now. The bug occurs in a state that is difficult to reproduce. 
I think we can easily trigger the bug and verify the patch with some new tests in fabtests. 
I will look at it later today.

Reply:
FYI, I managed to reproduce the bug with some manual fault injection in my code.
And I confirm the PR fixes the user-after-free detected by ASAN :+1: .